examples/chat/tui/commands.nim

```diff
@@ -27,6 +27,62 @@ const hashCharSet = {'#'}
 # `command` proc should implement the appropriate logic to deal with those
 # values
 
+# AddCustomToken ----------------------------------------------------------------
+
+proc help*(T: type AddCustomToken): HelpText =
+  let command = "addcustomtoken"
+  HelpText(command: command, aliases: aliased[command], parameters: @[
+    CommandParameter(name: "address", description: "Address of the " &
+      "custom token."),
+    CommandParameter(name: "name", description: "Name of " &
+      "the custom token."),
+    CommandParameter(name: "symbol", description: "Symbol of the " &
+      "custom token."),
+    CommandParameter(name: "color", description: "(Optional) Color (in hex) for " &
+      "custom token."),
+    CommandParameter(name: "decimals", description: "(Optional) Number of decimals to use for " &
+      "the custom token.")
+    ], description: "Creates a new custom token ")
+
+proc new*(T: type AddCustomToken, args: varargs[string]): T =
+  var
+    address = ""
+    name = ""
+    symbol = ""
+    color = ""
+    decimals = ""
+  if args.len > 2:
+    address = args[0]
+    name = args[1]
+    symbol = args[2]
+  if args.len > 4:
+    color = args[3]
+    decimals = args[4]
+
+  T(address: address, name: name, symbol: symbol, color: color, decimals: decimals)
+
+proc split*(T: type AddCustomToken, argsRaw: string): seq[string] =
+  argsRaw.split(" ")
+
+proc command*(self: ChatTUI, command: AddCustomToken) {.async, gcsafe,
+  nimcall.} =
+
+  try:
+    if command.address == "":
+      self.wprintFormatError(epochTime().int64,
+        "address cannot be empty, please provide an address.")
+    elif command.name == "":
+      self.wprintFormatError(epochTime().int64,
+        "name cannot be empty, please provide a name.")
+    elif command.symbol == "":
+      self.wprintFormatError(epochTime().int64,
+        "symbol cannot be empty, please provide a symbol.")
+    else:
```

Comment:
Because there are two optional parameters, we will never be able to know if the user was specifying `color` and not `decimals` or `decimals` and not `color`. So maybe we should provide some validation here, along the lines of "if color specified, check for valid hex string. if decimals specified, check for valid int". What do you think?

Reply:
Agree. Do you mean we'd check that if we have just 4 (not 5 params), we'd then check if the fourth parameter begins with `0x`, and if so, try to parse it as hex, otherwise try to parse it as int?

Reply:
Maybe we could try using `parseAddress` from `nim_status/conversions` with a try/catch, and if it fails, then we could consider it as invalid? Something like this:

```nim
  var addressParsed: Address
  try:
    addressParsed = address.parseAddress
  except Exception as e:
    # not valid, show validation error
```

Reply:
`addCustomToken` in `tasks.nim` does the address parsing too and throws the exception if it fails, as per your suggestion below.

Reply:
Right, but I think this is a separate issue - one is for validation (which helps to figure out which parameter the user was trying to specify),  and the other is for parsing the address to use in the nim-status client. One option would be to change the `AddCustomToken.address` type to `Address` and you won't need to worry about double-parsing.

Reply:
Add address/decimals parsing to `commands.nim`. I've attempted to pass more specific types (`Address` and uint for address and decimals), but in that case `client/tasks.nim/AddCustomToken` compilation fails with
```
nim-status/vendor/nim-web3/web3/ethtypes.nim(231, 25) Error: ambiguous call; both typetraits.distinctBase(T: typedesc) [declared in nim-status/vendor/nimbus-build-system/vendor/Nim/lib/pure/typetraits.nim(70, 6)] and sugar.distinctBase(T: typedesc) [declared in nim-status/vendor/nimbus-build-system/vendor/Nim/lib/pure/sugar.nim(164, 10)] match for: (FieldType)
```
Have no idea what this is about, so left it untyped for now.

Reply:
Maybe try instead of importing all of the sugar module, import only `=>`, ie:
```nim
import std/sugar #<== how it currently is
# to
from std/sugar import `=>`
```

Reply:
Thanks for the tip, this helped!